torch/csrc/jit/codegen/onednn/kernel.cpp

```diff
@@ -0,0 +1,258 @@
+    auto spec = inputSpecs_[i];
+    runInputs.push_back({spec.logical_tensor(), inputs[i].data_ptr()});
+  }
+  for (size_t i = 0; i < constantInputs_.size(); i++) {
+    // constantInputSpecs are placed after graphInputSpecs
+    auto constantInputSpecIdx = nGraphInputs_ + i;
+    auto constantInputSpec = inputSpecs_[constantInputSpecIdx];
+    runInputs.push_back(
+        {constantInputSpec.logical_tensor(), constantInputs_[i].data_ptr()});
+  }
+
+  for (size_t i = 0; i < nOutputs_; i++) {
+    auto spec = outputSpecs_[i];
+    auto opt = c10::TensorOptions(spec.aten_scalar_type()).device(device_);
+
+    auto outputId = spec.tid();
+    auto iter = inplacePairs_.find(outputId);
+    if (iter != inplacePairs_.end()) {
+      // output reuses one of input tensors
+      auto inputOffset = iter->second;
+      auto inputTensor = inputs[inputOffset];
+      outputs.push_back(inputTensor);
+      runOutputs.push_back({spec.logical_tensor(), inputTensor.data_ptr()});
+    } else if (spec.is_opaque()) {
+      auto tensor = empty_llga(spec, opt);
+      outputs.push_back(tensor);
+      runOutputs.push_back(llga_from_aten_tensor(tensor));
+    } else {
+      auto tensor = at::empty(spec.sizes(), opt);
+      outputs.push_back(tensor);
+      runOutputs.push_back({spec.logical_tensor(), tensor.data_ptr()});
+    }
+  }
+
+  return std::make_tuple(runInputs, runOutputs);
+}
+
+compiled_partition LlgaKernel::compile(const partition& partition) {
+  auto inputs = fmap(inputSpecs_, toLogicalTensor);
+  auto outputs = fmap(outputSpecs_, toLogicalTensor);
+  auto compilation = partition.compile(inputs, outputs, Engine::getEngine());
+
+  // Since layouts of opaque outputs would be known after compilation,
+  // we need to query them out from compilation and update outputSpecs
+  for (size_t i = 0; i < nOutputs_; i++) {
+    auto tid = outputSpecs_[i].tid();
+    outputSpecs_[i] = compilation.query_logical_tensor(tid);
+  }
+
+  // Build static mapping from output id to input offset
+  // in accordance with available inplace options
+  for (auto&& option : compilation.get_inplace_ports()) {
+    size_t inputId = option.first;
+    size_t outputId = option.second;
+    auto inputSpecIter =
+        std::find_if(inputSpecs_.begin(), inputSpecs_.end(), [&](auto& spec) {
+          return spec.tid() == inputId;
+        });
+    TORCH_CHECK(inputSpecIter != inputSpecs_.end(), "In-place input not found");
+    auto inputOffset = inputSpecIter - inputSpecs_.begin();
+    inplacePairs_[outputId] = inputOffset;
+  }
+
+  return compilation;
+}
+
+void LlgaKernel::run(Stack& stack) {
+  GRAPH_DEBUG("In ", debugName(), "\n");
+
+  // Grab input values from stack
+  auto stackInputs = last(stack, nGraphInputs_);
+  auto inputs = fmap(stackInputs, [&](const IValue& v) {
+    TORCH_CHECK(
+        v.isTensor(), "Stack values for LLGA partition must be Tensor type");
+    return v.toTensor();
+  });
+
+  lock_read();
+  if (is_initialized_) {
```

Comment:
one minor: just wondering if it's possible to use a standard mutex here, which is more cross-platform.

```
  if (!is_initialized_) {
    std::lock_guard<std::mutex> lock(rw_mutex_);
    if (!is_initialized_) {
      ...
      is_initialized_ = true;
    }
  }

  private:
    std::mutex rw_mutex_;
 ```

Reply:
Thanks for your inputs, @pinzhenx!
@chunyuan-w, I took another look & noticed that the custom `lock_read` & `lock_write` primitives are only being used here, so I'm confused why we need to use a read-write lock here?

The Reader using the read-lock doesn't seem to be reading anything -
```
lock_read();
  if (is_initialized_) {
    unlock_read();
  }
```

Reply:
Thanks! I noticed that the custom RW lock is only being used here, but it's not actually needed here,
so I'll get rid of it.

Reply:
Uh, example above is just describing an idea of portable code. Double checked lock like this isn’t robust actually. It should be further combined with std::atomic (even specify memory order on non-x86). Other than this hand written way, a modern while portable approach could be achieved by std::call_once here.

Reply:
Thanks, @pinzhenx. I'll take a look at what exactly the lock is being used for, and will make changes accordingly.

Also, I think atomics aren't required here for correctness because even if a reader were to read a stale value, it'd definitely read the correct one after acquiring the lock because there can be only one writer, which only writes after acquiring the lock. Atomics slow things down, so I'd rather not use it.

However, the current snippet has two differences from the previous version, due to which I was considering revising it anyway:
1. In the previous version, readers would've waited for the writer to finish writing. (I'll see what exactly the lock is being used for, and revisit this decision).
2. In this version, each reader would've to wait on `std::mutex`, which is a performance-bottleneck, so I'll use a RW mutex here.

Reply:
cc @jgong5